### ISSUE TITLE
Decouple shadows from `ATTENUATION` into its own `SHADOW_ATTENUATION` built-in in `spatial` shaders

### ISSUE THREAD

**Issue by Lauson1ex** [godotengine/godot-proposals#2905]
### Describe the project you are working on
Third-person, high-fidelity visuals 3d game.

### Describe the problem or limitation you are having in your project
In `spatial` shaders, there's a built-in `uniform` that couples distance, aperture AND shadows attenuation into one single `ATTENUATION` built-in. As per #2904, this is undesired and creates a slew of problems when creating custom light shaders.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
I propose shadows to be decoupled from distance/aperture attenuation, into its own built-in.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Introduce the `SHADOW_ATTENUATION` built-in to `spatial` shaders. This will only contain its namesake. The `ATTENUATION` built-in will now only contain range light attenuation information. Life is good!

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Introducing the `SHADOW_ATTENUATION` built-in has many other applications than just creating custom light shaders: users are able to know if a fragment is in shadow (0.0 is completely shadowed, 1.0 is completely lit, values in-between tell you that the fragment is only partially shadowed), or creating artistically-controlled colored shadows, for example.

### Is there a reason why this should be core and not an add-on in the asset library?
This is core.

**Comment by hidemat** [comment]
I agree with this. I've been trying to implement a toon shader in godot 4. Everything looks good until I enable attenuation.

Without ATTENUATION:
![image](https://user-images.githubusercontent.com/6559151/163509463-5b95202c-0221-4b88-8fe9-84439a81bffb.png)

With ATTENUATION:
![image](https://user-images.githubusercontent.com/6559151/163509536-cff39492-337b-421a-873e-9d528e2f2d83.png)

This is the shader code:
```
shader_type spatial;

//render_mode ambient_light_disabled;

uniform vec4 albedo : hint_color = vec4(1.0f);
uniform sampler2D albedo_texture : hint_albedo;
uniform bool clamp_diffuse_to_max = false;

uniform int cuts : hint_range(1, 8) = 3;
uniform float wrap : hint_range(-2.0f, 2.0f) = 0.0f;
uniform float steepness : hint_range(1.0f, 8.0f) = 1.0f;

uniform bool use_attenuation = true;

uniform bool use_specular = true;
uniform float specular_strength : hint_range(0.0f, 1.0f) = 1.0f;
uniform float specular_shininess : hint_range(0.0f, 32.0f) = 16.0f;
uniform sampler2D specular_map : hint_albedo;

uniform bool use_rim = true;
uniform float rim_width : hint_range(0.0f, 1.2) = 1.0f;
uniform vec4 rim_color : hint_color = vec4(1.0f);

uniform bool use_ramp = false;
uniform sampler2D ramp : hint_albedo;

uniform bool use_borders = false;
uniform float border_width = 0.01f;

varying vec3 vertex_pos;
varying vec3 normal;

float split_specular(float specular) {
	return step(0.5f, specular);
}

void vertex() {
	vertex_pos = VERTEX;
	normal = NORMAL;
}

void fragment() {
	ALBEDO = albedo.rgb * texture(albedo_texture, UV).rgb;
}

void light() {
	// Attenuation.
	float attenuation = 1.0f;
	if (use_attenuation) {
		attenuation = ATTENUATION;
	}
	
	// Diffuse lighting.
	float NdotL = dot(NORMAL, LIGHT);
	float diffuse_amount = NdotL + (attenuation - 1.0) + wrap;
	//float diffuse_amount = NdotL * attenuation + wrap;
	diffuse_amount *= steepness;
	float cuts_inv = 1.0f / float(cuts);
	float diffuse_stepped = clamp(diffuse_amount + mod(1.0f - diffuse_amount, cuts_inv), 0.0f, 1.0f);

	// Calculate borders.
	float border = 0.0f;
	if (use_borders) {
		float corr_border_width = length(cross(NORMAL, LIGHT)) * border_width * steepness;
		border = step(diffuse_stepped - corr_border_width, diffuse_amount)
				 - step(1.0 - corr_border_width, diffuse_amount);
	}
	
	// Apply diffuse result to different styles.
	vec3 diffuse = ALBEDO.rgb * LIGHT_COLOR / PI;
	if (use_ramp) {
		diffuse *= texture(ramp, vec2(diffuse_stepped * (1.0f - border), 0.0f)).rgb;
	} else {
		diffuse *= diffuse_stepped * (1.0f - border);
	}
	
	if (clamp_diffuse_to_max) {
		// Clamp diffuse to max for multiple light sources.
		DIFFUSE_LIGHT = max(DIFFUSE_LIGHT, diffuse);
	} else {
		DIFFUSE_LIGHT += diffuse;
	}
	
	// Specular lighting.
	if (use_specular) {
		vec3 H = normalize(LIGHT + VIEW);
		float NdotH = dot(NORMAL, H);
		float specular_amount = max(pow(NdotH, specular_shininess*specular_shininess), 0.0f)
							    * texture(specular_map, UV).r
								* attenuation;
		specular_amount = split_specular(specular_amount);
		SPECULAR_LIGHT += specular_strength * specular_amount * LIGHT_COLOR;
	}
	
	// Simple rim lighting.
	if (use_rim) {
		float rim_dot = 1.0 - dot(NORMAL, VIEW);
		float rim_light = rim_dot * NdotL;
		rim_light = smoothstep(rim_width - 0.01, rim_width + 0.01, rim_light);
		
		DIFFUSE_LIGHT += rim_light * rim_color.rgb * rim_color.a * LIGHT_COLOR;
	}
}
```

**Comment by miss-programgamer** [comment]
Hi, I am working on a toon shader that would benefit from this proposal, so I thought I would share the details of my roadblock in the hopes of motivating the Godot team to consider increasing the priority of implementing this feature.

My use case is a toon shader, not dissimilar to the one used in wind waker hd, where lights don't just have an effect on the front of objects, but also illuminate their backside slightly, giving them a softer and more pleasant look. This works fine for directional lights, but problems arise for lights that have distance attenuation, like spotlights and omnilights. In short, shadows cast onto surfaces by other objects need to still be illuminated slightly so that they match the backside, but distance attenuation needs to be gradual all the way to zero, so as to avoid an unpleasant pop when an object exists the range of the light. This is impossible to implement without distance and shadow attenuation being implemented as separate values, since shadows and distance have different effects on the overall shading of the object.

Another way of implementing this would be to keep the existing ATTENUATION builtin working as it currently does, and to add two more builtins, SHADOW_ATTENUATION and DISTANCE_ATTENUATION, that way no existing code would be broken by this addition and could be implemented in a minor version update.

I hope you take another look at this proposal and consider implementing it soon, looking forward to updates on this!

**Comment by Calinou** [comment]
I can't find where the `ATTENUATION` built-in is defined in the codebase in the `master` branch.

There might be a technical reason for the two values being put together, especially performance (since optimizing shaders to reduce VGPR occupancy is a priority).

**Comment by clayjohn** [comment]
They were combined during the big push to optimize VGPR usage. Of course, based on community feedback, I think we are seeing that the tradeoff may not have been worthwhile. 

I think we need to investigate how to expose these separately without increasing VGPR usage. 

Attenuation is needed for the built in shaders, but perhaps we could expose ``ATTENUATION`` ``SHADOW_ATTENUATION`` and ``DISTANCE_ATTENUATION`` and then only send the value that needs to be used. For example if the user shader only uses ``SHADOW_ATTENUATION`` then both ``ATTENUATION`` and ``DISTANCE_ATTENUATION`` would be ignored. 

While this would slightly increase the complexity of the base shader code, it sounds like it would be worthwhile.

**Comment by novaplusplus** [comment]
Bit of a bump, but I have been working with some shader lighting code and this would be very useful for me. I would definitely like to see this or something like it brought around.

**Comment by Calinou** [comment]
@novaplusplus Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by lostminds** [comment]
I think this would be a great addition, either as SHADOW_ATTENUATION or perhaps SHADOW_OPACITY, to link it to the shadow opacity property of the light that will be part of it. 

Regardless of naming, I think this could also make it easier to fix the issues with the "Shadow to opacity" setting being broken in StandardMaterial3D https://github.com/godotengine/godot/issues/62257

**Comment by QbieShay** [comment]
Additional usecase: I'm making an interactive snow shader. I want self-shadowing from the displacement, but I want also to add a scatter component for when there is a buried object that emits light (omnilight). currently there's no way for me to have object inside "scatter" and objects outside receive shadows.

No shadow
![image](https://github.com/godotengine/godot-proposals/assets/7917475/89f7fd5c-f887-4933-93b8-dbab3b6fb284)
With shadow
![image](https://github.com/godotengine/godot-proposals/assets/7917475/f646c94a-5e90-4215-aca4-5537f60f2c87)
Shadow i'd like to preserve but can't
![image](https://github.com/godotengine/godot-proposals/assets/7917475/e885e2ed-db3b-45b3-b525-ee3a1d126b78)
